### PR TITLE
etcd+itests: use itest generated ports for embedded etcd and add etcd itests to travis CI

### PR DESCRIPTION
**Pull request by bhandras**
This PR adds optional user defined ports for the embedded etcd instance in order to fix the issue when parallel itests try to open the same ports. The last commit adds support for CI testing of our etcd stack (on bitcoind).

### CODE REVIEW

**Review by guggero — COMMENTED** [review]
Code looks good to me!
Congratulations on a successful itest run (looks like only flakes on the non-etcd tests) :tada:

Interesting that the itest with `etcd` passed, but with new errors detected in the logs:
```
<time> [ERR] CNCT: ChainArbitrator: unable to mark ChannelPoint(<chan_point>) fully closed: no closed channel for chan_point=<chan_point> found
<time> [ERR] CNCT: ChannelArbitrator(<chan_point>): unable to advance state: no closed channel for chan_point=<chan_point> found
<time> [ERR] CNCT: Unable to advance state: no closed channel for chan_point=<chan_point> found
<time> [ERR] CNCT: unable to mark channel resolved: no closed channel for chan_point=<chan_point> found
```

Could this indicate that come channel state isn't being retrieved correctly? Perhaps a bug in the `etcd` related code?

File: channeldb/kvdb/config.go [hunk 1]
```diff
@@ -36,7 +36,11 @@ type BoltConfig struct {
 
 // EtcdConfig holds etcd configuration.
 type EtcdConfig struct {
-	Embedded bool `long:"embedded" description:"Use embedded etcd instance instead of the external one."`
+	Embedded bool `long:"embedded" description:"Use embedded etcd instance instead of the external one. Note: use for testing only."`
+
+	EmbeddedClientPort int `long:"embedded_client_port" description:"Client port to use for the embedded instance. Note: use for testing only."`
```
[the remark this answers is not in the document]
Comment (bhandras) [reply channeldb/kvdb/config.go]: done

File: sample-lnd.conf [hunk 2]
```diff
@@ -984,6 +984,12 @@ litecoin.node=ltcd
 ; Useful for testing.
 ; db.etcd.embedded=false
 
+; If non zero, LND will use this as client port for the embedded etcd instance.
+; db.etcd.embedded_client_port=1234
+
+; If non zero, LND will use this as peer port for the embedded etcd instance.
+; db.etcd.embedded_peer_port = 1235
```
[the remark this answers is not in the document]
Comment (bhandras) [reply sample-lnd.conf]: done

**Comment by bhandras** [comment]
> Code looks good to me!
> Congratulations on a successful itest run (looks like only flakes on the non-etcd tests) tada
> 
> Interesting that the itest with `etcd` passed, but with new errors detected in the logs:
> 
> ```
> <time> [ERR] CNCT: ChainArbitrator: unable to mark ChannelPoint(<chan_point>) fully closed: no closed channel for chan_point=<chan_point> found
> <time> [ERR] CNCT: ChannelArbitrator(<chan_point>): unable to advance state: no closed channel for chan_point=<chan_point> found
> <time> [ERR] CNCT: Unable to advance state: no closed channel for chan_point=<chan_point> found
> <time> [ERR] CNCT: unable to mark channel resolved: no closed channel for chan_point=<chan_point> found
> ```
> 
> Could this indicate that come channel state isn't being retrieved correctly? Perhaps a bug in the `etcd` related code?

I believe these errors were a result of `kvdb.Batch` calling into the `walletdb` implementation for etcd too. This is now fixed (first commit)

**Comment by bhandras** [comment]
> > Code looks good to me!
> > Congratulations on a successful itest run (looks like only flakes on the non-etcd tests) tada
> > Interesting that the itest with `etcd` passed, but with new errors detected in the logs:
> > ```
> > <time> [ERR] CNCT: ChainArbitrator: unable to mark ChannelPoint(<chan_point>) fully closed: no closed channel for chan_point=<chan_point> found
> > <time> [ERR] CNCT: ChannelArbitrator(<chan_point>): unable to advance state: no closed channel for chan_point=<chan_point> found
> > <time> [ERR] CNCT: Unable to advance state: no closed channel for chan_point=<chan_point> found
> > <time> [ERR] CNCT: unable to mark channel resolved: no closed channel for chan_point=<chan_point> found
> > ```
> > 
> > 
> > Could this indicate that come channel state isn't being retrieved correctly? Perhaps a bug in the `etcd` related code?
> 
> I believe these errors were a result of `kvdb.Batch` calling into the `walletdb` implementation for etcd too. This is now fixed (first commit)

Seems like the issue was rather that we attempted to mark a breached channel fully closed while the close summary was not yet stored in the DB.

**Review by guggero — APPROVED** [review]
Nice, LGTM :100:

File: contractcourt/chain_watcher.go [hunk 3]
```diff
@@ -1182,6 +1169,19 @@ func (c *chainWatcher) dispatchContractBreach(spendEvent *chainntnfs.SpendDetail
 	log.Infof("Breached channel=%v marked pending-closed",
 		c.cfg.chanState.FundingOutpoint)
 
+	// With the event processed and channel closed, we'll now notify all
+	// subscribers of the event.
+	c.Lock()
+	for _, sub := range c.clientSubscriptions {
```
Comment (Roasbeef) [inline contractcourt/chain_watcher.go:1175]: Excellent catch! 

I think this might be the culprit of some of the flakes we've seen in this area as well.

**Review by Roasbeef — APPROVED** [review]
LGTM 🍘

**Comment by Roasbeef** [comment]
Major milestone to have the `etcd` tests reliably passing!

**Pull request merged**